lib/marked.js

```diff
@@ -7,8 +7,223 @@
+
+    // Not sure what all is happening here.
+    var highlight = opt.highlight,
+        tokens,
+        pending,
+        i = 0;
+
+    // Run the lexer, if it fails; bail early by calling the callback.
+    // Would like to move this somewhere else. Make most of the returns toward the end; or,
+    // more obvious.
+    try {
+      tokens = Lexer.lex(src, opt)
+    } catch (e) {
+      return callback(e);
+    }
+
+    // Internal block for promise-type behavior??
+    var done = function(err) {
+      if (err) {
+        opt.highlight = highlight;
+        return callback(err);
+      }
+
+      var out;
+
+      try {
+        out = Parser.parse(tokens, opt);
+      } catch (e) {
+        err = e;
+      }
+
+      opt.highlight = highlight;
+
+      return err
+        ? callback(err)
+        : callback(null, out);
+    };
+
+    // Highlight option check, not set or less than three bail
+    //
+    // TODO: Don't some highlighter plugins allow for `js`, effectively being less than three
+    //       despite wanting to have the highlighter run??
+    if (!highlight || highlight.length < 3) {
+      return done();
+    }
+
+    delete opt.highlight;
+
+    pending = tokens.length;
+    if (!pending) return done();
+
+    for (; i < tokens.length; i++) {
+      (function(token) {
+        if (token.type !== 'code') {
+          return --pending || done();
+        }
+        return highlight(token.text, token.lang, function(err, code) {
+          if (err) return done(err);
+          if (code == null || code === token.text) {
+            return --pending || done();
+          }
+          token.text = code;
+          token.escaped = true;
+          --pending || done();
+        });
+      })(tokens[i]);
+    }
+
+    return;
+  }
+
+  // marked without options or callback...not tested, can't seem to be moved to pass tests.
+  try {
+    if (opt) opt = merge({}, marked.defaults, opt);
+    return Parser.parse(Lexer.lex(src, opt), opt);
+  } catch (e) {
+    e.message += '\nPlease report this to https://github.com/chjj/marked.';
+    if ((opt || marked.defaults).silent) {
+      return '<p>An error occurred:</p><pre>'
+        + escape(e.message + '', true)
+        + '</pre>';
+    }
+    throw e;
+  }
+}
+
+/**
+ * TODO: Not sure how to test this with our testing harness.
+ *
+ * @throws Error If src is null, undefined, or not a string.
+ */
+marked.receivedValidSource = function(src) {
+  // throw error in case of non string input
+  if (typeof src === 'undefined' || src === null) {
+    return new Error('marked(): input parameter is undefined or null');
+  }
+  if (typeof src !== 'string') {
+    return new Error('marked(): input parameter is of type '
+      + Object.prototype.toString.call(src) + ', string expected');
+  }
+}
+
+/**
+ * TODO: Not sure how to test
+ *
+ * @param  src String
+ * @return boolean
+ */
+marked.receivedOnlySource = function(src, opt, callback) {
+  return (marked.receivedCallback(opt, callback));
```

Comment:
Should be return `!(marked.receivedCallback(opt, callback));` only source means we did *not* receive those other things.